### PR TITLE
[astro] Fix offsets if "earliest" and "latest" is null or default "00:00"

### LINKED ISSUE #19694: [astro] Offsets not applied to DateTime channels and events unless "earliest" and "latest" are configured, and not "00:00"
## Expected Behavior

Offsets (both positive and negative) should work even without configuration of "earliest" and "latest".

## Current Behavior

Positive offsets are only applied if "latest" is explicitly configured and not set to "00:00".
Same thing for negative offsets, but for "earliest".

Got introduced by #19310 

## Possible Solution

Currently investigating

## Steps to Reproduce (for Bugs)

1. Configure an Astro Sun Thing with either
```yaml
  - id: set#start
    channelTypeUID: astro:start
    label: Start Time
    description: The start time of the event
    configuration:
      offset: 60
  - id: set#end
    channelTypeUID: astro:end
    label: End Time
    description: The end time of the event
    configuration: 
      offset: -60
```
or (which gets applied if using the configuration panel in MainUI)
```yaml
  - id: set#start
    channelTypeUID: astro:start
    label: Start Time
    description: The start time of the event
    configuration:
      offset: 60
      earliest: 00:00
      latest: 00:00
  - id: set#end
    channelTypeUID: astro:end
    label: End Time
    description: The end time of the event
    configuration:
      offset: -60
      earliest: 00:00
      latest: 00:00
```

2. Connect Items to the channels
3. The offsets are not applied.
4. To make it work the configuration needs to be
```yaml
  - id: set#start
    channelTypeUID: astro:start
    label: Start Time
    description: The start time of the event
    configuration:
      offset: 60
      earliest: 00:00
      latest: 23:59
  - id: set#end
    channelTypeUID: astro:end
    label: End Time
    description: The end time of the event
    configuration:
      offset: -60
      earliest: 00:01
      latest: 00:00
```
## Context

Previously working offsets stopped working after an update. Especially requiring "earliest" to be configured to "00:01" (or later) instead of "00:00" is unintuitive.

## Your Environment

* Version used: (e.g., openHAB and add-on versions)
* Environment name and version (e.g. Chrome 76, Java 8, Node.js 12.9, ...):
* Operating System and version (desktop or mobile, Windows 10, Raspbian Buster, ...):
```yaml
runtimeInfo:
  version: 5.1.0.M2
  buildString: Milestone Build
locale: en-SE
systemInfo:
  configFolder: /etc/openhab
  userdataFolder: /var/lib/openhab
  logFolder: /var/log/openhab
  javaVersion: 21.0.9
  javaVendor: Azul Systems, Inc.
  javaVendorVersion: Zulu21.46+19-CA
  osName: Linux
  osVersion: 6.1.0-40-amd64
  osArchitecture: amd64
  availableProcessors: 4
  freeMemory: 396006912
  totalMemory: 1556086784
  uptime: 325874
  startLevel: 100
addons:
  - automation-jrubyscripting
  - binding-astro
  - binding-gardena
  - binding-melcloud
  - binding-myuplink
  - binding-network
  - binding-pushover
  - binding-rfxcom
  - binding-spotify
  - binding-systeminfo
  - binding-tibber
  - binding-unifi
  - binding-verisure
  - binding-zigbee
  - binding-zwave
  - misc-openhabcloud
  - persistence-influxdb
  - persistence-mapdb
  - transformation-map
  - transformation-scale
clientInfo:
  device:
    ios: false
    android: false
    androidChrome: false
    desktop: true
    iphone: false
    ipod: false
    ipad: false
    edge: false
    ie: false
    firefox: false
    macos: false
    windows: true
    cordova: false
    phonegap: false
    electron: false
    nwjs: false
    webView: false
    webview: false
    standalone: false
    os: windows
    pixelRatio: 2
    prefersColorScheme: dark
  isSecureContext: true
  locationbarVisible: true
  menubarVisible: true
  navigator:
    cookieEnabled: true
    deviceMemory: 8
    hardwareConcurrency: 20
    language: sv-SE
    languages:
      - sv-SE
      - en-SE
      - en-GB
      - en-US
      - sv
      - en
    onLine: true
    platform: Win32
  screen:
    width: 1440
    height: 900
    colorDepth: 24
  support:
    touch: false
    pointerEvents: true
    observer: true
    passiveListener: true
    gestures: false
    intersectionObserver: true
  themeOptions:
    dark: dark
    filled: true
    pageTransitionAnimation: default
    bars: light
    homeNavbar: default
    homeBackground: default
    expandableCardAnimation: default
    blocklyRenderer: null
  userAgent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML,
    like Gecko) Chrome/142.0.0.0 Safari/537.36
timestamp: 2025-11-23T10:19:24.791Z

### PR DESCRIPTION
**Pull request by pacive**
Fixes #19694

### CODE REVIEW

**Comment by jlaur** [comment]
Regression of #19310. @Nadahar - perhaps you'd like to have a look?

**Comment by Nadahar** [comment]
Sorry for being a bit late, I just want to figure out why this happened in the first place. #19310 wasn't meant to change behavior at all, just prevent bugs, so I'm wondering why it ended up changing behavior.

@pacive I've you've already figured out what "went wrong", please give me a hint so I can save some time 😉

**Comment by Nadahar** [comment]
The changes to these two methods in #19310 seem to have been:
```diff
@@ -205,11 +213,11 @@ public class DateTimeUtils {
     }
 
     public static Calendar getAdjustedEarliest(Calendar cal, AstroChannelConfig config) {
-        return adjustTime(cal, getMinutesFromTime(config.earliest));
+        return config.earliest == null ? cal : adjustTime(cal, getMinutesFromTime(config.earliest));
     }
 
     public static Calendar getAdjustedLatest(Calendar cal, AstroChannelConfig config) {
-        return adjustTime(cal, getMinutesFromTime(config.latest));
+        return config.latest == null ? cal : adjustTime(cal, getMinutesFromTime(config.latest));
     }
```
So, I guess it boils down to what `getMinutesFromTime()` used to do when it was handed `null`?

**Comment by pacive** [comment]
The specific changes that altered the behaviour was the changes to `public static Calendar applyConfig(Calendar cal, AstroChannelConfig config)` in `DateTimeUtils.java`. The calls to `getAdjustedEarliest` and `getAdjustedLatest` were changed to use the original `cal` object instead of the `cCal` object that had the offset already applied.

When I just changed those lines back, some tests failed however, so I figured there was some reason for this change, and I changed the implementations of `getAdjustedEarliest` and `getAdjustedLatest` to check the configuration and only adjust the time if not null and not "00:00"

**Comment by Nadahar** [comment]
The memories of this is quite vague now, unfortunately, but I think that the reason to use `cal` instead of `cCal` for `getAdjustedEarliest()` and `getAdjustedLatest()` was that if using the value with the offset would sometimes cross over into the wrong date and return "earliest" for the previous day or "latest" for the next day. The original offset calculation ignored time zone, so it would fail if applied to a calendar instance with a different timezone than what was the default on the computer, which I think partly masked the "date jump bug".

~~But, it seems that the "fix" might have been overly simplistic, basically ignoring the offset if `earliest` or `latest` are set.~~

edit: The above sentence is wrong, it doesn't ignore offset if `earliest` or `latest` are set, but it intends to let them get the final word. However, there is a flaw there, because either `cCal.before(cEarliest)` or `cCal.after(cLatest)` will always be true if there is an offset, making one of them return instead of the offset value.

I think this is the "real" bug.

**Comment by Nadahar** [comment]
Your implementation applies "earliest" and "latest" even if they aren't set (in setting it to the start or end of the current date). I can't wrap my head around why that is the correct thing to do. If these aren't set, shouldn't `cal` just be returned unaltered?

**Comment by Nadahar** [comment]
It also strikes me that this is about how the logic is meant to be. Basically, should earliest/latest or offset "have the last word"? Should an offset be applied which can take the time outside earliest/latest, or should those be absolutes. I guess my thinking at the time was that the limits are absolutes, so if the offset take you outside these, the offset becomes irrelevant.

**Comment by Nadahar** [comment]
I'd say that this isn't the way to address it. The problem is in `applyConfig()`, and that's where it needs to be solved. `getAdjustedEarliest()` and `getAdjustedLatest()` are public API (although that might be of limited importance in a binding) and are used by other methods. This basically completely redefines what they do.

Instead, `applyConfig` should have its bug fixed. I'm just not entirely sure what's the best way to do that yet.

**Comment by pacive** [comment]
After thinking a bit more I agree. The way I see it it should be:
1. Latest/earliest set: offset should be applied, but not in a way that goes outside the bounds.
2. Latest/earliest NOT set: set earliest/latest to start/end of the day; goto 1

Need to look at my fix again to see if that's the behaviour it produces.

**Comment by Nadahar** [comment]
> 2\. Latest/earliest NOT set: set earliest/latest to start/end of the day; goto 1

I'm not sure about this. My interpretation is that if the limits aren't set, no limits should be applied. That way, you can cross midnight using an offset, if you so desire. If you don't want to be able to cross midnight, you can apply that as a limit.

edit: However, there's a problem defining midnight as "earliest", because `0` is treated as "not defined", so I don't think that limit would be respected.

**Comment by pacive** [comment]
Yes, it currently only works if earliest is set to 00:01. It would be easier if only `null` needed to be handled, but unfortunately, if configuring an offset in MainUI, both earliest and latest are set to 00:00 by default and that gets saved even if those fields aren't touched, so for ease of use those need to be treated the same. Otherwise the only way would be to go into the code tab and remove the parameters.

**Comment by Nadahar** [comment]
Ah, I didn't know about MainUI behavior here, but that explains this "strange behavior".

Edit: That actually causes another bug: `createCalendarForToday()` will return "an arbitrary time" if called with `0, 0`.

**Comment by pacive** [comment]
I guess we need some input from @openhab/maintainers on what the expected behaviour should be WRT crossing date boundaries, I can see both ways as equally valid.

**Comment by Nadahar** [comment]
> I guess we need some input from @openhab/maintainers on what the expected behaviour should be WRT crossing date boundaries, I can see both ways as equally valid.

The question is if anybody actually has a stance on this. I think historic behavior might be the most relevant guide to use.

**Comment by Nadahar** [comment]
@pacive I've come up with this:
```diff
 .../binding/astro/internal/util/DateTimeUtils.java | 53 +++++++++++++++++-----
 1 file changed, 42 insertions(+), 11 deletions(-)

diff --git a/bundles/org.openhab.binding.astro/src/main/java/org/openhab/binding/astro/internal/util/DateTimeUtils.java b/bundles/org.openhab.binding.astro/src/main/java/org/openhab/binding/astro/internal/util/DateTimeUtils.java
index 3c6f3d89e2..81d69d2065 100644
--- a/bundles/org.openhab.binding.astro/src/main/java/org/openhab/binding/astro/internal/util/DateTimeUtils.java
+++ b/bundles/org.openhab.binding.astro/src/main/java/org/openhab/binding/astro/internal/util/DateTimeUtils.java
@@ -213,11 +213,15 @@ public class DateTimeUtils {
     }
 
     public static Calendar getAdjustedEarliest(Calendar cal, AstroChannelConfig config) {
-        return config.earliest == null ? cal : adjustTime(cal, getMinutesFromTime(config.earliest));
+        int minutes = getMinutesFromTime(config.earliest);
+        // MainUI sets earliest to 00:00 if unconfigured, which is why zero must be treated as such
+        return minutes > 0 ? adjustTime(cal, minutes) : cal;
     }
 
     public static Calendar getAdjustedLatest(Calendar cal, AstroChannelConfig config) {
-        return config.latest == null ? cal : adjustTime(cal, getMinutesFromTime(config.latest));
+        int minutes = getMinutesFromTime(config.latest);
+        // MainUI sets latest to 00:00 if unconfigured, which is why zero must be treated as such
+        return minutes > 0 ? adjustTime(cal, minutes) : cal;
     }
 
     /**
@@ -230,20 +234,45 @@ public class DateTimeUtils {
             cCal.add(Calendar.MINUTE, config.offset);
         }
 
-        Calendar cEarliest = getAdjustedEarliest(cal, config);
-        if (cCal.before(cEarliest)) {
-            return cEarliest;
+        int minutes = getMinutesFromTime(config.earliest);
+        Calendar threshold, actual;
+        // MainUI sets earliest to 00:00 if unconfigured, which is why zero must be treated as such
+        if (minutes > 0) {
+            if ((threshold = truncateToMidnight(cal)).equals(actual = truncateToMidnight(cCal))) {
+                // Same day
+                Calendar cEarliest = getAdjustedEarliest(cCal, config);
+                if (cCal.before(cEarliest)) {
+                    return cEarliest;
+                }
+            } else {
+                // Previous or next day
+                if (actual.before(threshold)) {
+                    return getAdjustedEarliest(threshold, config);
+                }
+            }
         }
-        Calendar cLatest = getAdjustedLatest(cal, config);
-        if (cCal.after(cLatest)) {
-            return cLatest;
+        minutes = getMinutesFromTime(config.latest);
+        // MainUI sets latest to 00:00 if unconfigured, which is why zero must be treated as such
+        if (minutes > 0) {
+            if ((threshold = endOfDayDate(cal)).equals(actual = endOfDayDate(cCal))) {
+                // Same day
+                Calendar cLatest = getAdjustedLatest(cCal, config);
+                if (cCal.after(cLatest)) {
+                    return cLatest;
+                }
+            } else {
+                // Previous or next day
+                if (actual.after(threshold)) {
+                    return getAdjustedLatest(threshold, config);
+                }
+            }
         }
 
         return cCal;
     }
 
     private static Calendar adjustTime(Calendar cal, int minutes) {
-        if (minutes > 0) {
+        if (minutes >= 0) {
             Calendar cTime = truncateToMidnight(cal);
             cTime.add(Calendar.MINUTE, minutes);
             return cTime;
@@ -256,7 +285,9 @@ public class DateTimeUtils {
     }
 
     /**
-     * Parses a HH:MM string and returns the minutes.
+     * Parses a HH:MM string and returns hours and minutes in minutes.
+     *
+     * @return The number of minutes from midnight, or {@code -1} if the string couldn't be parsed.
      */
     private static int getMinutesFromTime(@Nullable String configTime) {
         if (configTime != null) {
@@ -279,6 +310,6 @@ public class DateTimeUtils {
 
             }
         }
-        return 0;
+        return -1;
     }
 }
```
It's not fully thought through or tested, but it's at least the gist of the way I think it should be handled. It fails with your tests, since your tests assumes that `getAdjustedEarliest()` and `getAdjustedLatest` returns the thresholds of the day if unconfigured, with I would argue doesn't match "adjust" which I interpret as "only change if outside".

**Comment by pacive** [comment]
Thanks, will test it out

**Comment by Nadahar** [comment]
@pacive I just pushed it here, to save you some copy/paste: https://github.com/Nadahar/openhab-addons/tree/astro-limits-regression

This commit to be specific: https://github.com/Nadahar/openhab-addons/commit/becb5fd3b02829b67543faf65eb91ded193016bc

**Comment by pacive** [comment]
Not sure if it's just my tests that are the problem, the assert on line 171 fails first, which is:
```java
assertEquals(fixedStart.getTime(), startNullResultStart.getTime());
```
`fixedStart` is defined as:
```
fixedStart = DateTimeUtils.getAdjustedEarliest(pointInTime, config); // will return 00:00
fixedStart.add(Calendar.MINUTE, 93);
```
`startNullResultStart` is defined by
```
startNullResult = Job.adjustRangeToConfig(startNull, config, TIME_ZONE, Locale.ROOT);
```
Which is what should be tested. However, the error says
```
JobTest.adjustRangeToConfigForceTest:171 expected: <Mon Nov 24 10:32:59 CET 2025> but was: <Sun Nov 23 23:01:48 CET 2025>
```
The time for `startNullResultStart` changes every time I run the test, so even if `fixedStart` was defined correctly, the test would never succeed.

**Comment by pacive** [comment]
Hmm, in this section:
```java
        Calendar threshold, actual;
        // MainUI sets earliest to 00:00 if unconfigured, which is why zero must be treated as such
        if (minutes > 0) {
            if ((threshold = truncateToMidnight(cal)).equals(actual = truncateToMidnight(cCal))) {
                // Same day
                Calendar cEarliest = getAdjustedEarliest(cCal, config);
                if (cCal.before(cEarliest)) {
                    return cEarliest;
                }
            } else {
                // Previous or next day
                if (actual.before(threshold)) {
                    return getAdjustedEarliest(threshold, config);
                }
            }
        }
```
`threshold` and `actual` doesn't get initialized if minutes <= 0 (which it it if config.earliest is null.

**Comment by Nadahar** [comment]
> `threshold` and `actual` doesn't get initialized if minutes <= 0 (which it it if config.earliest is null.

No, but they aren't used if that's the case either...?

This is a "pattern" I often use, to define the variables outside the scope where they are used, if I need the same types several places. The idea is merely to not allocate more variables on the stack for no reason, and to reuse them instead. I'm not saying that it makes a big difference, but I like to take into consideration what the consequences of the code are. I typically never initialize variables inside a loop either - it might be that the compiler will "optimize" this anyway, but I can't know that, so why not just make it so in the first place.

The Java compiler won't actually compile if you use a variable that might not have been initialized (via some path), so I think it's a pretty risk-free approach.

**Comment by Nadahar** [comment]
> The time for `startNullResultStart` changes every time I run the test, so even if `fixedStart` was defined correctly, the test would never succeed.

It's no wonder why it keeps changing, it's using the current time the test is run (with timezone applied). The question is how this is supposed to work, I don't know what the logic has been behind this code:

```java
    static Range adjustRangeToConfig(Range range, AstroChannelConfig config, TimeZone zone, Locale locale) {
        Calendar start = range.getStart();
        Calendar end = range.getEnd();

        if (config.forceEvent && start == null) {
            start = getAdjustedEarliest(Calendar.getInstance(zone, locale), config);
        }
```
`startNull` has a `null` start (hence the name), and when exposed to `adjustRangeToConfig()` while `config.forceEvent` is `true`, `Calendar.getInstance(zone, locale)` is called. That will return a "now instance", in the specified timezone. Adjusting that to "earliest" is quite pointless, at least for a test, since the result will depend on when the test is run. Also, in this case, `earliest` is `null`, so no adjustment will be performed. As a result, you will have a "start" value that corresponds to the time when the test is being run.

I'm not quite sure exactly what "force" is supposed to do, except that it seems to "forbid" `null`. The decision to just substitute `null` with "now" is questionable in my opinion, I'm not sure why that has been chosen.

**Comment by pacive** [comment]
Sorry, must have been tires and misread the else branch to be for the outer if.

I noticed also that it uses getInstance, but what I find strange is that all the other test cases work anyway, and my tests was at least deterministic before I applied your patch.

My guess is that "force" is meant to fire the event at the earliest/latest time configured even if the event would not happen (e.g., sunrise when there's midnight sun). This was one of the reasons I felt setting the bounds to start/end of day was a reasonable approach.

**Comment by pacive** [comment]
I guess in practice this would mean that both start and end events would fire the moment the job is scheduled by the binding (+offset), which should be close to midnight. Not sure how negative offsets would work here, but I guess they would be fired immediately.

**Comment by Nadahar** [comment]
> I noticed also that it uses get instance, but what I find strange is that all the other test cases work anyway, and my tests was at least deterministic before I applied your patch.

That isn't so strange, because you made the "adjust" methods default to "the extremes" if no limits were defined. But, this wasn't the behavior in the past, and the "adjust" methods are also used by other code (and are public), which is why I think it's a bad idea to do.

That said, we can make it do what we want, my problem is that I don't understand "the intention" - what exactly does it mean to "force" a range.

I never dug into all the details in the logic of the binding, I just tried to apply bug fixes to the logic that was already there. Unfortunately, that's not always possible, because fixing one bug just reveals another. But, it's still not clear to me what the consequences of setting the threshold when "forcing" an open-ended range is. If that means that the range should start at 0:00 and end at 23:59:59.999 - that's easy enough to achieve - I just need to understand properly what the goal is, to be able to tell what "the right behavior" should be.

**Comment by pacive** [comment]
I think the intention of "force" is to force execute the event at the configured earliest/latest even if no event should have occured (such as when there's no sunrise). The question is how should this apply when no bounds are configured? 

In the original it seemed to fire immediately when created, since no bounds would be applied. On the rare days where an event such as "night" have a start but no end (sun don't go up again until >1 day later), this would have the weird effect that "end" is triggered immediately, but "start" later at the intended time.

If that is the intention or just an oversight I don't know. Makes it more difficult to test for though.

**Comment by Nadahar** [comment]
According to [the documentation](https://www.openhab.org/addons/bindings/astro/#:~:text=Force), you are right:

> Force event: For each trigger channel and start, end datetime value, you can force the earliest, latest time of the day, when the event is actually not taking place (e.g. astronomic dawn during summer in Sweden) e.g sun#astroDawn earliest=6:00, latest=20:00 forceEvent=true

With that definitions, it sounds like `adjustRangeToConfig()` should be modified to apply the first and last point in the day for `null` endpoints if `force` is true. It still doesn't really make sense to me, because we still have to use "now" to determine what day whose extremes to find, so when looking at this event from another date, its endpoints will be "nonsense". But, maybe that's OK.

Alternatively, we could use the date from the defined end (one of the ends in a range should be defined, shouldn't it?). That way, it wouldn't depend on the current date. Whatever we do here, it feels like we're defining something new, I don't think this behavior has ever been properly defined and that it has just "happen to work" in the use cases most commonly seen.

**Comment by Nadahar** [comment]
Based on the definition of "force", I've attempted an implementation that I hope will be somewhat close. I'm still not really "onboard" with why you'd want to do this, but regardless:

```diff
 .../java/org/openhab/binding/astro/internal/job/Job.java | 16 +++++++++++-----
 1 file changed, 11 insertions(+), 5 deletions(-)

diff --git a/bundles/org.openhab.binding.astro/src/main/java/org/openhab/binding/astro/internal/job/Job.java b/bundles/org.openhab.binding.astro/src/main/java/org/openhab/binding/astro/internal/job/Job.java
index b0790d17bc..f48e95f0e6 100644
--- a/bundles/org.openhab.binding.astro/src/main/java/org/openhab/binding/astro/internal/job/Job.java
+++ b/bundles/org.openhab.binding.astro/src/main/java/org/openhab/binding/astro/internal/job/Job.java
@@ -136,11 +136,17 @@ public interface Job extends SchedulerRunnable, Runnable {
         Calendar start = range.getStart();
         Calendar end = range.getEnd();
 
-        if (config.forceEvent && start == null) {
-            start = getAdjustedEarliest(Calendar.getInstance(zone, locale), config);
-        }
-        if (config.forceEvent && end == null) {
-            end = getAdjustedLatest(Calendar.getInstance(zone, locale), config);
+        if (config.forceEvent) {
+            Calendar reference = start != null ? start : end;
+            if (reference == null) {
+                reference = Calendar.getInstance(zone, locale);
+            }
+            if (start == null) {
+                start = getAdjustedEarliest(truncateToMidnight(reference), config);
+            }
+            if (end == null) {
+                end = getAdjustedLatest(endOfDayDate(reference), config);
+            }
         }
 
         // depending on the location and configuration you might not have a valid range for day/night, so skip the
```
This can also be found here: https://github.com/Nadahar/openhab-addons/commit/3385f0cfab53d54ccabde1164ba42ecbd118d489

The idea here is to try to get "the most reasonable information" from the little we have. So, if one of the endpoints is `null` and `forceEvent` is `true`, the day from the other endpoint is used. If both are `null`, the "now" date is used (because we *have no information to use*). This is what I call "the reference". The reference is then taken either to the very start or to the very end of the day before "earliest" and "latest" are applied - which should make it align with them *if they are defined*. Otherwise, the event will start/end when the reference date starts/ends.

Does this make sense?

Additionally, it's clear that more tests are needed. The original tests still pass with this "redefinition" of `adjustRangeToConfig()`, it seems that very few corner cases are covered by tests, which is why behavior changes aren't picked up by the tests.

**Comment by Nadahar** [comment]
I've created some more tests to try to get better coverage for some of the "Util" methods: https://github.com/Nadahar/openhab-addons/commit/74d024df22bd78a04659e630448bb015ef75c35b

**Comment by Nadahar** [comment]
@jlaur @pacive I'm a bit unsure of how to proceed. I think my suggested modifications takes care of the immediate issue. But, too much is untested here, the tests should cover more aspects to "fix" behavior in place. It's hard to know if something unexpected has changed as the result of doing modifications when the tests are this "forgiving". So, should we try to address that now or not?

**Comment by pacive** [comment]
I've been thinking about the ranges and "force" setting a bit more, and I'm not sure I understand the logic behind how it's implemented. An example just to demonstrate some weirdness:

Let's say I want to turn on some lights when the sun goes down, so I create a rule that triggers on the set START event. But I would like the lights to turn on earliest at 16:00 and 21:00 at the latest, so I configure those as well. Now it happens that I live in the far north of Sweden, so during summer, the sun is up all day around and there is no sunset. This causes the range [START..END] that defines the triggers to be [null..null]. If I enable the "force" setting, when the `adjustRangeToConfig` function is called it will set the range to [16:00..21:00], which means that the START event will be triggered at 16:00. If I instead choose to trigger on the END event, the opposite will happen during winter, as soon as there no longer is a sunset, the rule will be triggered at 21:00, when the day before triggered at 16:00 (since sunset would be close to noon, so it would be adjusted to "earliest").

If I instead would opt to use the "daylight#event" channel, I would of course only be able to use the END event (the START would be for turning off the lights in the morning). Then I can only set the "latest" parameter to restrict the time of the events. Daylight END would become closer to noon during the fall, until there is no daylight anymore, and at that day the event will suddenly jump to trigger at 21:00.

I honestly can't see a valid use case for forcing the events even when null that wouldn't cause similar issues.

Edit: Note that for the DateTime channels, there are two channels #start and #end, which can be configured separately, but the #event channel sets the bounds for both START and END with one config. One solution could be to let the config for #start/#end affect #event as well (this would also reduce some common confusion that pops up in the forum from time to time about why the settings don't work). But this would be a breaking change that many without these problems would need to adapt to.

**Comment by Nadahar** [comment]
You don't have to sell me on the (lack of) logic with the `null` endpoints. As I said above, I don't quite understand what we're supposed to do with those. Instead of the "force" setting, a "default" setting (where the user could specify a time that would be used when `null`) seems more useful to me.

But, we're here to try to restore what was, as a bugfix. I think it might be smart to do some rethinking of some of this, but for now, I think we need to find something that makes it "work" (as good or bad as in the past) without any reconfiguration.

I don't think what we have done now matches what was done in the past, where I believe that "now" was substituted for `null`. I'm not able to say exactly what that means, because the result will depend on exactly when the evaluation runs.

**Comment by pacive** [comment]
The event ranges are created and events scheduled at 30s past midnight local time [here](https://github.com/openhab/openhab-addons/blob/0e1b4481a55d8e5cf6d80f000f2b1e09c047e397/bundles/org.openhab.binding.astro/src/main/java/org/openhab/binding/astro/internal/handler/AstroThingHandler.java#L210). So using the original implementation they would always be scheduled the same day. If no earliest/latest is configured (but "force" is), since null is treated as 0 (minutes after midnight) both START and END will have be 00:00 the current day. However, it seems when the [scheduling happens](https://github.com/openhab/openhab-addons/blob/0e1b4481a55d8e5cf6d80f000f2b1e09c047e397/bundles/org.openhab.binding.astro/src/main/java/org/openhab/binding/astro/internal/job/Job.java#L53) it only does so if the event is after the current time, which means that the events will not be scheduled to run.

This means that having the MainUI default "00:00" would be equal to not having any configuration (= null), so to be certain the event will fire "earliest" would need to be set to "00:01".

**Comment by Nadahar** [comment]
> However, it seems when the [scheduling happens](https://github.com/openhab/openhab-addons/blob/0e1b4481a55d8e5cf6d80f000f2b1e09c047e397/bundles/org.openhab.binding.astro/src/main/java/org/openhab/binding/astro/internal/job/Job.java#L53) it only does so if the event is after the current time, which means that the events will not be scheduled to run.

Yes - so is what we have now in any way "worse" than what has been? Some features exist but are hardly used, maybe "force" is one of those? I mean, many people might use it, not knowing what it does, but they rarely/never have endpoints with `null`, so it has no actual effect...

**Comment by pacive** [comment]
Probably not, so the only thing we need to worry about is to ensure the offset is applied, without bounds if not earliest/latest is configured, and restrict it to those bounds if they are. If the offset causes the time to cross midnight, the event will not be scheduled anyway, so null and "00:00" should be completely equal.

**Comment by Nadahar** [comment]
> so the only thing we need to worry about is to ensure the offset is applied, without bounds if not earliest/latest is configured, and restrict it to those bounds if they are.

I believe that is working with the code I have suggested. The real bug here was that the offset was never applied, because either the "before" or the "after" check would always trigger. That isn't the case anymore, and I've written tests that validate this behavior.

**Comment by Nadahar** [comment]
Maybe it's easier if I create a PR with the code I have? Or you can just use it here, it doesn't matter to me, but I'm thinking that it might be difficult "to get the overview" of what I've done as it is now?

**Comment by pacive** [comment]
You can do the PR, and I'll just close this one, seems easiest!

**Comment by pacive** [comment]
Superseded by #19717
